### PR TITLE
Rename CUB test variants to be less confusing and add a `CUB_TEST_VARIANT` CMake option to control them.

### PR DESCRIPTION
**Pull request by brycelelbach**
Historically, each CUB test has had up to three different test variants, controlled by macro defines.
 
* *"thorough" or no variant:* Comprehensive unit testing which typically covers all builtin test types, all scenarios, and has support for repeating the tests multiple times (via a `--repeat` flag).
* *"quicker"*: Minimal unit testing intend to compile and run in a shorter period of time and may only cover a subset of test types and scenarios.
* *"quick"*: Performance comparisons of CUB primitives with other libraries (usually Thrust, but also NPP in some places) and against theoretical hardware capabilities (memory bandwidth, etc). When CUB was initially released, Thrust had its own CUDA backend, which CUB aimed to compete with. The Thrust backend was replaced with a CUB-based backend because CUB was successful in that competition; as such, a number of these comparisons are less meaningful than they once were. However, the performance comparisons relative to theoretical hardware capabilities remain valuable.

The names "quicker" and "quick" are a bit confusing for the uninitiated; you might think that they are two different levels of thoroughness, however in reality one is for benchmarking and one is for minimal testing.

My guess is that "quick" (the benchmarks) were added first, and then later the "quicker" variants were added. At the time "quicker" was added, it's likely no one realized it would be a bit confusing.

This PR renames "quicker" to "minimal" and "quick" to "benchmark".

It also adds a `CUB_TEST_VARIANT` CMake option which can be used to control which test variants are built.
* `ALL` (default and current behavior): Build all available variants.
* `THOROUGH`: If a test has multiple variants, only build the thorough variant.
* `BENCHMARK`: If a test has multiple variants, only build the benchmark variant 
* Add a `CUB_TEST_VARIANT` option which controls which test variants are built.

Originally I was going to call this option `CUB_TEST_WORKLOAD`, but I didn't want to confuse the workload terminology with the terminology used by gpuCI and Thrust multi-config.

### CODE REVIEW

**Comment by brycelelbach** [comment]
run tests

File: CMakeLists.txt [hunk 1]
```diff
@@ -48,8 +48,13 @@ endif()
 
 option(CUB_ENABLE_HEADER_TESTING "Test that all public headers compile." ON)
 option(CUB_ENABLE_TESTING "Build CUB testing suite." ON)
+set(CUB_TEST_VARIANT "ALL" CACHE STRING "Specify which CUB test variants should be built (options: ALL, THOROUGH, BENCHMARK, MINIMAL).")
```
Comment (allisonvacanti) [inline CMakeLists.txt:51]: Can we make this a series of booleans

```
CUB_TEST_VARIANT_THOROUGH
CUB_TEST_VARIANT_MINIMAL
CUB_TEST_VARIANT_BENCHMARK
```

that default to `ON`? It would allow us to just turn off `THOROUGH` variants but leave still have some testing + benchmarks.

File: CMakeLists.txt [hunk 2]
```diff
@@ -48,8 +48,13 @@ endif()
...
 option(CUB_ENABLE_EXAMPLES "Build CUB examples." ON)
 
+if (NOT CUB_TEST_VARIANT MATCHES "^(ALL|THOROUGH|BENCHMARK|MINIMAL)$")
```
Comment (allisonvacanti) [inline CMakeLists.txt:54]: This is moot if we switch to booleans, but FYI there's a better way to do this:

```
  set_property(CACHE CUB_TEST_VARIANT PROPERTY STRINGS
    ALL THOROUGH BENCHMARK MINIMAL
  )
```

This will populate dropdown menus, etc in CMake GUIs.

File: test/Makefile [hunk 3]
```diff
@@ -43,10 +43,10 @@ include ../common.mk
 
 # Testing mode option (quick/thorough)
 ifeq ($(quickertest), 1)
-	NVCCFLAGS += -DQUICKER_TEST
+	NVCCFLAGS += -DMINIMAL_TEST
```
Comment (allisonvacanti) [inline test/Makefile:46]: This doesn't match the `#ifdef`s, should be `CUB_TEST_MINIMAL`.

File: test/Makefile [hunk 4]
```diff
@@ -43,10 +43,10 @@ include ../common.mk
...
 	TEST_SUFFIX = quicker
 else ifeq ($(quicktest), 1)
-	NVCCFLAGS += -DQUICK_TEST
+	NVCCFLAGS += -DBENCHMARK
```
Comment (allisonvacanti) [inline test/Makefile:49]: Same, should be `CUB_TEST_BENCHMARK`

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -60,27 +60,27 @@ function(cub_add_test target_name_var test_name test_src cub_target)
   )
 endfunction()
 
-# Sets HAS_QUICK_VARIANT / HAS_QUICKER_VARIANT / NO_VARIANTS to True/False in
+# Sets HAS_BENCHMARK_VARIANT / HAS_MINIMAL_VARIANT / NO_VARIANTS to True/False in
 # the calling scope.
 # Used to detect variants of unit tests depending on whether a source file
-# contains the strings "QUICK_TEST" or "QUICKER_TEST".
+# contains the strings "BENCHMARK" or "MINIMAL_TEST".
```
Comment (allisonvacanti) [inline test/CMakeLists.txt:66]: These should check for the full strings `CUB_TEST_BENCHMARK`, `CUB_TEST_MINIMAL` to avoid false positives.

Checking for `MINIMAL_TEST` won't match the new `#ifdef`s (`CUB_TEST_MINIMAL`).

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -60,27 +60,27 @@ function(cub_add_test target_name_var test_name test_src cub_target)
...
 function(cub_check_for_test_variants src)
   file(READ "${src}" data)
 
-  string(FIND "${data}" "QUICK_TEST" quick_loc)
-  set(HAS_QUICK_VARIANT False PARENT_SCOPE)
-  if (NOT quick_loc EQUAL -1)
-    set(HAS_QUICK_VARIANT True PARENT_SCOPE)
+  string(FIND "${data}" "BENCHMARK" benchmark_loc)
```
Comment (allisonvacanti) [inline test/CMakeLists.txt:70]: `CUB_TEST_BENCHMARK`

File: test/CMakeLists.txt [hunk 7]
```diff
@@ -60,27 +60,27 @@ function(cub_add_test target_name_var test_name test_src cub_target)
...
+  set(HAS_BENCHMARK_VARIANT False PARENT_SCOPE)
+  if (NOT benchmark_loc EQUAL -1)
+    set(HAS_BENCHMARK_VARIANT True PARENT_SCOPE)
   endif()
 
-  string(FIND "${data}" "QUICKER_TEST" quicker_loc)
-  set(HAS_QUICKER_VARIANT False PARENT_SCOPE)
-  if (NOT quicker_loc EQUAL -1)
-    set(HAS_QUICKER_VARIANT True PARENT_SCOPE)
+  string(FIND "${data}" "MINIMAL_TEST" minimal_loc)
```
Comment (allisonvacanti) [inline test/CMakeLists.txt:76]: `CUB_TEST_MINIMAL`

File: test/CMakeLists.txt [hunk 8]
```diff
@@ -91,40 +91,50 @@ foreach (test_src IN LISTS test_srcs)
   get_filename_component(test_name "${test_src}" NAME_WE)
   string(REGEX REPLACE "^test_" "" test_name "${test_name}")
 
-  # Some tests change behavior based on whether the compiler defs QUICK_TEST
-  # and/or QUICKER_TEST are defined. Detect these and build variants for each
+  # Some tests change behavior based on whether the compiler defs BENCHMARK
```
Comment (allisonvacanti) [inline test/CMakeLists.txt:94]: Strings need updating.

File: test/CMakeLists.txt [hunk 9]
```diff
@@ -91,40 +91,50 @@ foreach (test_src IN LISTS test_srcs)
...
+  # and/or MINIMAL_TEST are defined. Detect these and build variants for each
   # configuration:
   cub_check_for_test_variants("${test_src}")
 
   foreach(cub_target IN LISTS CUB_TARGETS)
     if (NO_VARIANTS)
       # Only one version of this test.
       cub_add_test(test_target ${test_name} "${test_src}" ${cub_target})
+    elseif (NOT CUB_TEST_VARIANT STREQUAL "ALL")
+      string(TOLOWER CUB_TEST_VARIANT lowercase_cub_test_variant)
```
Comment (allisonvacanti) [inline test/CMakeLists.txt:104]: This can be removed and rolled into the logic below if we switch to booleans options:
1. wrap the `thorough` logic in an `if (CUB_TEST_VARIANT_THOROUGH)`
2. update the `HAS_` checks to `if (CUB_TEST_VARIANT_XXX AND HAS_XXX_VARIANT)`

File: test/CMakeLists.txt [hunk 10]
```diff
@@ -91,40 +91,50 @@ foreach (test_src IN LISTS test_srcs)
...
+      target_compile_definitions(${test_target_${lowercase_cub_test_variant}}
+                                 PRIVATE CUB_TEST_${CUB_TEST_VARIANT})
     else()
       # By default (no flags), the "thorough" version of the test is built:
       cub_add_test(test_target_thorough
         ${test_name}.thorough
         "${test_src}"
         ${cub_target}
       )
+      target_compile_definitions(${test_target_thorough} PRIVATE CUB_TEST_THOROUGH)
```
Comment (allisonvacanti) [inline test/CMakeLists.txt:119]: Nothing defines this, is this addition intentional?

Comment (allisonvacanti) [reply test/CMakeLists.txt:119]: If we add this here, it should be added to the Makefile, too.

File: test/test_device_radix_sort.cu [hunk 11]
```diff
@@ -1220,14 +1220,14 @@ int main(int argc, char** argv)
...
-#ifdef QUICKER_TEST
+#ifdef CUB_TEST_MINIMAL
 
     enum {
         IS_DESCENDING   = false
     };
 
     // Compile/run basic CUB test
-    if (num_items < 0)      num_items       = 48000000;
+    if (num_items < 0)      num_items       = 24000000;
```
Comment (allisonvacanti) [inline test/test_device_radix_sort.cu:1230]: This change looks unrelated. Was it intentional?

**Review by allisonvacanti — CHANGES_REQUESTED** [review]
The explanation and overall change is great.

I'd like to tweak the implementation to simplify some of the logic and provide more flexibility over which tests get generated, see inline comments.

File: CMakeLists.txt (hunk 1 above)
Comment (brycelelbach) [reply CMakeLists.txt:51]: Done.

File: CMakeLists.txt (hunk 2 above)
Comment (brycelelbach) [reply CMakeLists.txt:54]: Done.

File: test/Makefile (hunk 3 above)
Comment (brycelelbach) [reply test/Makefile:46]: Can we just delete the makefiles at this point? Who uses them? We don't.

File: test/CMakeLists.txt (hunk 5 above)
Comment (brycelelbach) [reply test/CMakeLists.txt:66]: Good catch.

File: test/CMakeLists.txt (hunk 6 above)
Comment (brycelelbach) [reply test/CMakeLists.txt:70]: Fixed

File: test/CMakeLists.txt (hunk 7 above)
Comment (brycelelbach) [reply test/CMakeLists.txt:76]: Fixed.

File: test/CMakeLists.txt (hunk 8 above)
Comment (brycelelbach) [reply test/CMakeLists.txt:94]: Fixed.

File: test/CMakeLists.txt (hunk 9 above)
Comment (brycelelbach) [reply test/CMakeLists.txt:104]: Fixed in a slightly different way.

File: test/test_device_radix_sort.cu (hunk 11 above)
Comment (brycelelbach) [reply test/test_device_radix_sort.cu:1230]: This is intentional; that test just takes too long right now.

File: test/CMakeLists.txt (hunk 10 above)
Comment (brycelelbach) [reply test/CMakeLists.txt:119]: Done. I just added it for completeness.

File: test/Makefile (hunk 3 above)
Comment (allisonvacanti) [reply test/Makefile]: Burn them 🔥🔥🔥

**Review by allisonvacanti — APPROVED** [review]
Looks good, approved.

Only major issue I have is that I'd still like to leave the suffixes on the test names for the single variant builds. It will make our lives easier when we need to debug CI problems and bug reports. What do you think?

File: test/CMakeLists.txt [hunk 12]
```diff
@@ -60,71 +60,97 @@ function(cub_add_test target_name_var test_name test_src cub_target)
...
+  # Some tests change behavior based on whether the compiler defs BENCHMARK
+  # and/or MINIMAL_TEST are defined. Detect these and build variants for each
   # configuration:
   cub_check_for_test_variants("${test_src}")
 
   foreach(cub_target IN LISTS CUB_TARGETS)
     if (NO_VARIANTS)
       # Only one version of this test.
       cub_add_test(test_target ${test_name} "${test_src}" ${cub_target})
+    elseif (single_test_variant_option)
```
[the remark this answers is not in the document]
Comment (brycelelbach) [reply test/CMakeLists.txt:117]: That's probably fine, I'll yank it out.

File: test/CMakeLists.txt (hunk 8 above)
Comment (brycelelbach) [reply test/CMakeLists.txt:94]: They are now.

File: CMakeLists.txt [hunk 13]
```diff
@@ -48,6 +48,9 @@ endif()
 
 option(CUB_ENABLE_HEADER_TESTING "Test that all public headers compile." ON)
 option(CUB_ENABLE_TESTING "Build CUB testing suite." ON)
+option(CUB_TEST_VARIANT_THOROUGH "Build CUB thorough test variants." ON)
```
[the remark this answers is not in the document]
Comment (brycelelbach) [reply CMakeLists.txt]: Done.
